src/ast.scm

```diff
@@ -523,6 +523,21 @@
   (and (if one (length= e 3) (length> e 2))
        (eq? (car e) 'meta) (memq (cadr e) '(nospecialize specialize))))
 
+(define (meta? e)
+  (and (length> e 1) (eq? (car e) 'meta)))
+
+(define (method-meta-sym? x)
+  (memq x '(inline noinline aggressive_constprop no_constprop propagate_inbounds)))
+
+(define (propagate-method-meta e)
+  `(meta ,@(filter (lambda (x)
+                     (or (method-meta-sym? x)
+                         (and (pair? x) (eq? (car x) 'purity))))
```

Comment:
For purity we can only keep the ones that are true of keyword sorters. Safer not to propagate them at all. But I think the following apply to keyword sorters:
- consistent
- effect_free
- terminates_globally
- notaskstate
But please check me on that!

Reply:
I think keyword sorters are usually:
- `:consistent`
- `:effect_free`
- `:terminates_locally` (`:terminates_globally` requires that of the body function itself)
- `:notaskstate`

but they are not `:nothrow` (since they can throw any of `TypeError`/`UndefKeywordError`/`MethodError`).

Well, I'd argue it is okay to propagate the effects assumed by `Base.@assume_effects` to keyword sorters, since the point of using the macro to override the conclusion that is otherwise derived by the effect analysis.
If we write
```julia
Base.@assume_effects :total specific(A::Type, B::Type; v::Number=1) =
   v * one(ccall(:jl_type_morespecific, Cint, (Any, Any), A, B) ≠ 0 ? A : B)
```
then IMO we'd like to assume both `specific(Int, Integer)` and `specific(Int, Integer; v=42)` are `:total` (and therefore `:nothrow` also)?